google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/ReadWriteTransaction.java

```diff
@@ -0,0 +1,1283 @@
+    Preconditions.checkNotNull(updates);
+    final List<Statement> updateStatements = new LinkedList<>();
+    for (ParsedStatement update : updates) {
+      Preconditions.checkArgument(
+          update.isUpdate(),
+          "Statement is not an update statement: " + update.getSqlWithoutComments());
+      updateStatements.add(update.getStatement());
+    }
+    checkValidTransaction();
+    try {
+      if (retryAbortsInternally) {
+        return asyncExecuteStatement(
+            executeBatchUpdateStatement,
+            new Callable<long[]>() {
+              @Override
+              public long[] call() throws Exception {
+                return runWithRetry(
+                    new Callable<long[]>() {
+                      @Override
+                      public long[] call() throws Exception {
+                        try {
+                          getStatementExecutor()
+                              .invokeInterceptors(
+                                  executeBatchUpdateStatement,
+                                  StatementExecutionStep.EXECUTE_STATEMENT,
+                                  ReadWriteTransaction.this);
+                          long[] updateCounts = txContext.batchUpdate(updateStatements);
+                          createAndAddRetriableBatchUpdate(updateStatements, updateCounts);
+                          return updateCounts;
+                        } catch (AbortedException e) {
+                          throw e;
+                        } catch (SpannerException e) {
+                          createAndAddFailedBatchUpdate(e, updateStatements);
+                          throw e;
+                        }
+                      }
+                    });
+              }
+            },
+            InterceptorsUsage
+                .IGNORE_INTERCEPTORS); // ignore interceptors here as they are invoked in the
+                                       // Callable.
+      } else {
+        return asyncExecuteStatement(
+            executeBatchUpdateStatement,
+            new Callable<long[]>() {
+              @Override
+              public long[] call() throws Exception {
+                return txContext.batchUpdate(updateStatements);
+              }
+            });
+      }
+    } catch (SpannerException e) {
+      handlePossibleInvalidatingException(e);
+      throw e;
+    }
+  }
+
+  @Override
+  public void write(Mutation mutation) {
+    Preconditions.checkNotNull(mutation);
+    checkValidTransaction();
+    mutations.add(mutation);
+  }
+
+  @Override
+  public void write(Iterable<Mutation> mutations) {
+    Preconditions.checkNotNull(mutations);
+    checkValidTransaction();
+    for (Mutation mutation : mutations) {
+      this.mutations.add(checkNotNull(mutation));
+    }
+  }
+
+  /** Create a pseudo COMMIT statement to allow it to be cancelled or time out. */
+  private final ParsedStatement commitStatement =
```

Comment:
sorry, could you explain what this does?

Reply:
I've added the following javadoc to the field:

```java
  /**
   * Create a COMMIT statement to use with the {@link #commit()} method to allow it to be
   * cancelled, time out or retried.
   *
   * {@link ReadWriteTransaction} uses the generic methods
   * {@link #executeAsync(ParsedStatement, Callable)} and {@link #runWithRetry(Callable)} to
   * allow statements to be cancelled, to timeout and to be retried. These methods require a
   * {@link ParsedStatement} as input. When the {@link #commit()} method is called directly, we do
   * not have a {@link ParsedStatement}, and the method uses this statement instead in order to use
   * the same logic as the other statements.
   */
```